crates/lang/codegen/src/generator/storage_item.rs

```diff
@@ -0,0 +1,247 @@
+                        &salt,
+                        i,
+                        field,
+                    )
+                })
+                .collect();
+
+            let fields = match variant.fields {
+                Fields::Named(_) => quote! { { #(#fields),* } },
+                Fields::Unnamed(_) => quote! { ( #(#fields),* ) },
+                Fields::Unit => quote! {},
+            };
+
+            quote! {
+                #(#attrs)*
+                #variant_ident #fields #discriminant
+            }
+        });
+
+        quote! {
+            #vis enum #enum_ident #generics {
+                #(#variants),*
+            }
+        }
+    }
+
+    fn generate_union(&self, union_item: DataUnion) -> TokenStream2 {
+        let item = self.item;
+        let union_ident = item.ident();
+        let vis = item.vis();
+        let generics = item.generics();
+        let salt = item.salt();
+
+        let fields = union_item
+            .fields
+            .named
+            .iter()
+            .enumerate()
+            .map(|(i, field)| {
+                convert_into_storage_field(union_ident, None, &salt, i, field)
+            });
+
+        quote! {
+            #vis union #union_ident #generics {
+                #(#fields),*
+            }
+        }
+    }
+
+    fn generate_type_check(&self) -> TokenStream2 {
+        let fields = self
+            .item
+            .all_used_types()
+            .into_iter()
+            .enumerate()
+            .map(|(i, ty)| {
+                let field_name = format_ident!("field_{}", i);
+                let span = ty.span();
+                quote_spanned!(span=>
+                    #field_name: #ty
+                )
+            });
+        let generics = self.item.generics();
+        let salt = self.item.salt();
+
+        quote! {
+            const _: () = {
+                struct Check #generics {
+                    salt: #salt,
+                    #(#fields),*
+                }
+            };
+        }
+    }
+}
+
+fn convert_into_storage_field(
+    struct_ident: &Ident,
+    variant_ident: Option<&syn::Ident>,
+    salt: &TokenStream,
+    index: usize,
+    field: &Field,
+) -> Field {
+    let field_name = if let Some(field_ident) = &field.ident {
+        field_ident.to_string()
+    } else {
+        index.to_string()
+    };
+
+    let variant_name = if let Some(variant_ident) = variant_ident {
+        variant_ident.to_string()
+    } else {
+        "".to_string()
+    };
+
+    let key = ink_primitives::KeyComposer::compute_key(
+        struct_ident.to_string().as_str(),
+        variant_name.as_str(),
+        field_name.as_str(),
+    );
+
+    let mut new_field = field.clone();
+    let ty = field.ty.clone();
+    let span = field.span();
+    new_field.ty = parse2(quote_spanned!(span =>
+        <#ty as ::ink_storage::traits::AutoItem<
+            ::ink_storage::traits::ManualKey<#key, #salt>,
+        >>::Type
+    ))
+    .unwrap();
```

Comment:
Could we get an error output a la 

```
.unwrap_or_else(|error| ::core::panic!("converting into storage field failed: {}", error))
```

here? Maybe with the span?

Reply:
We can add span, but it requires changing the return type to `Result`.

I prefer not to do that because we can't get an error here=) `ty` already was parsed early, so it is a valid type. `key` and `salt` are autogenerated, so they also are valid.